### PR TITLE
Isolate just embed

### PR DESCRIPTION
**Pull request by prozacchiwawa**
Isolate just the (embed-file ...) forms from the larger compile-file and embed-file pr.

### CODE REVIEW

File: src/compiler/comptypes.rs [hunk 1]
```diff
@@ -228,6 +228,14 @@ pub enum HelperForm {
     Defun(bool, DefunData),
 }
 
+/// To what purpose is the file included.
```
Comment (aqk) [inline src/compiler/comptypes.rs:231]: This comment could be more descriptive.

File: src/compiler/preprocessor.rs [hunk 2]
```diff
@@ -30,6 +52,64 @@ pub fn process_include(
...
+            &HashMap::new(),
+            loc.clone(),
+            &decode_string(&content),
+        )
+        .map_err(run_to_compile_err)?,
+        IncludeProcessType::SExpression => {
+            let parsed = parse_sexp(Srcloc::start(&full_name), content.iter().copied())
+                .map_err(|e| CompileErr(e.0, e.1))?;
+            if parsed.len() != 1 {
+                return Err(CompileErr(loc, format!("More than one form in {fname}")));
```
Comment (aqk) [inline src/compiler/preprocessor.rs]: Does this error message need to include the 0 length case?

File: src/compiler/preprocessor.rs [hunk 3]
```diff
@@ -67,27 +148,63 @@ fn process_pp_form(
...
-                [SExp::Atom(kw, inc), SExp::QuotedString(nl, _, fname)] => {
-                    if "include".as_bytes().to_vec() == *inc {
-                        return Ok(Some(IncludeDesc {
-                            kw: kw.clone(),
-                            nl: nl.clone(),
-                            name: fname.clone(),
-                        }));
+
+                [SExp::Atom(kl, embed_file), SExp::Atom(_, name), SExp::Atom(_, kind), SExp::Atom(nl, fname)] => {
+                    if embed_file == b"embed-file" {
```
Comment (aqk) [inline src/compiler/preprocessor.rs:175]: TODO: Add the syntax rules for `embed-file`:
must exactly one of "hex", "bin", "sexp" be specified?

File: src/tests/classic/run.rs [hunk 4]
```diff
@@ -534,6 +534,123 @@ fn test_treehash_constant_embedded_modern_loop() {
...
+fn test_embed_file_5() {
+    let program = do_basic_run(&vec![
+        "run".to_string(),
+        "-i".to_string(),
+        "resources/tests".to_string(),
+        "(mod () (embed-file testsexp sexp embed.sexp) testsexp)".to_string(),
+    ])
+    .trim()
+    .to_string();
+    let run_result = do_basic_brun(&vec!["brun".to_string(), program, "()".to_string()])
```
Comment (aqk) [inline src/tests/classic/run.rs]: Can use `brun -n`

File: src/tests/classic/run.rs [hunk 5]
```diff
@@ -534,6 +534,123 @@ fn test_treehash_constant_embedded_modern_loop() {
...
+    assert_eq!(run_result, "(lsh 24 25)");
+}
+
+#[test]
+fn test_embed_file_7() {
+    let program = do_basic_run(&vec![
+        "run".to_string(),
+        "-i".to_string(),
+        "resources/tests".to_string(),
+        "(mod () (embed-file hello bin hello.bin) hello)".to_string(),
```
Comment (aqk) [inline src/tests/classic/run.rs]: Could make the name of the file different than the symbol name :)

**Comment by aqk** [comment]
Test matrix:

(Modern, Class)
(hex, sexp, bin)
(file exists, file does not exist)
(error thrown during processing, or not)

File: src/tests/classic/run.rs (hunk 4 above)
Comment (prozacchiwawa) [reply src/tests/classic/run.rs]: thanks

File: src/tests/classic/run.rs (hunk 5 above)
Comment (prozacchiwawa) [reply src/tests/classic/run.rs]: as in not hello.bin?

File: src/tests/classic/run.rs [hunk 6]
```diff
@@ -534,6 +534,128 @@ fn test_treehash_constant_embedded_modern_loop() {
...
+        .to_string();
+    let run_result = do_basic_brun(&vec!["brun".to_string(), program, "()".to_string()])
+        .trim()
+        .to_string();
+    assert_eq!(
+        run_result,
+        "0x9595c9df90075148eb06860365df33584b75bff782a510c6cd4883a419833d50"
+    );
+}
+
```
Comment (aqk) [inline src/tests/classic/run.rs:658]: What do you thing about adding a test here that includes the "embed-file", but not in function-call context. eg.

```
(concat "hello " "embed-file")
```

**Comment by aqk** [comment]
Another suggestion for a test: test the preprocessor on different patterns of interleaved code and embed-file use:

* embed file A
* embed file B
* filler code
* use embedded-data B
* use embedded-data A

eg.
```
(embed-file testhex hex hex-embed-01.hex)
(include sha256tree.clib)
(embed-file testsexp sexp embed.sexp)
; do something with embedded data
```

Some things these tests could encode:
* What order are different forms allowed to be listed in the program (`<code>`, `include`, `embed-file`, ...?
* Is there any unintended persistent state in the parser or pre-processor that can interact?

Let's discuss if you think these are needed / useful.

**Review by aqk — APPROVED** [review]

**Pull request merged**